FHEM/00_SIGNALduino.pm

```diff
@@ -1087,9 +1093,23 @@ SIGNALduino_ResetDevice($)
 {
   my ($hash) = @_;
   my $name = $hash->{NAME};
-
-  $hash->{logMethod}->($name, 3, "$name reset"); 
+  my $hardware = AttrVal($name,"hardware","");
+  $hash->{logMethod}->($name, 3, "$name/reset: $hardware"); 
   DevIo_CloseDev($hash);
+	if ($hardware eq "radinoCC1101" && $^O eq 'linux') {
+		# The reset is triggered when the Micro's virtual (CDC) serial / COM port is opened at 1200 baud and then closed.
+		# When this happens, the processor will reset, breaking the USB connection to the computer (meaning that the virtual serial / COM port will disappear).
+		# After the processor resets, the bootloader starts, remaining active for about 8 seconds.
+		# The bootloader can also be initiated by pressing the reset button on the Micro.
+		# Note that when the board first powers up, it will jump straight to the user sketch, if present, rather than initiating the bootloader.	
+		my $dev = $hash->{DeviceName};
+		my $baudrate;
+		($dev, $baudrate) = split("@", $dev);
+		$hash->{logMethod}->($name, 3, "$hash->{TYPE} $name/reset: forcing special reset for $hardware on $dev");
+		# Mit dem Linux-Kommando 'stty' die Port-Einstellungen setzen
+		system("stty -F $dev ospeed 1200 ispeed 1200");
+		sleep(1);	# ohne funktioniert es nicht
```

Comment:
Da pausieren wir ja den kompletten fhem Prozess für eine komplette Sekunde.
Ich denke darüber nach, ob wir das auf das sleep verzichten können und alternativ mit einem internalTimer auskommen könnten.

Reply:
Ich hatte es schon geahnt, das dieser Einwand kommen würde. Ich bemerke allerdings bei mir kein Lahmlegen von FHEM.
Kann es sein, das die interne Funktion "CommandSleep" das schon abfängt?

Reply:
nein, das ist ein perl Sleep. Das geht nicht mehr durch den FHEM Interpreter.

Ich hab da aber schon eine Idee. Bin gerade am umstellen des codes.
Vieleicht fällt mir auch noch ein, wie wir das ganze testen können

Reply:
Du hast es sicherlich gesehen, ich habe einen Änderungsvorschlag für das sleep:

https://github.com/elektron-bbs/RFFHEM/pull/1

Reply:
Nein, hatte ich noch nicht bemerkt. Ich werde es testen.
Was mir auf Anhieb auffällt: Das "sleep(1)" in der "SIGNALduino_ResetDevice" stört dich wohl nicht?

Reply:
Doch, aber ich wollte erst Mal einen Teil lösen um mit der Erkenntnis dann weiter zu machen

Reply:
Das tut es noch nicht:
2019.10.07 17:50:24 3: SIGNALduino sduinoACM/flash: forcing special reset for radinoCC1101 on /dev/ttyACM0
Undefined subroutine &main::internalTimer called at ./FHEM/00_SIGNALduino.pm line 643.

Reply:
Hat sich erledigt, war nur ein Tippfehler:
internalTimer - InternalTimer
Jetzt funktioniert es.

Reply:
Ich habe den Branch dev-r34-radino-reset-2 noch angepasst.

Die Reset Funktion wäre damit auch nicht blockierend. 
Tests habe ich für die verschiedenen Optionen auch erstellt

Reply:
OK, auch das funktioniert wunderprächtig :-)
Ich habe übrigens beide Varianten (Flash und Reset) vorsichtshalber mit dem radinoCC1101 und einem nano328 am gleichen System getestet. Hier funktioniert alles.

Wie bekommen wir jetzt den Pull request und den Branch dev-r34-radino-reset-2 zusammen?

Was mir noch einfällt: Das automatische Flashen beim Radino funktioniert nur, wenn der Radino mit /dev/ttyACMx definiert ist. Mit einer Definition in der Art /dev/serial/by-id muss der User den Radino weiterhin manuell in den Bootloader-Mode versetzen. Vieleicht sollte dazu noch ein Hinweis in die Commandref.

Reply:
Was für eine Fehlermeldung kommt denn, wenn der by-id Variante gewählt wird?
Ist das vielleicht eher eine Timing sache, da der Bootloader ja unter einer anderen ID läuft?

Reply:
Die Fehlermeldung von avrdude lautet sinngemäß: "Port nicht gefunden".
Wenn der Bootloader aktiv ist, ist die Bezeichnung anders:
ls -l /dev/serial/by-id
Normalbetrieb:
lrwxrwxrwx 1 root root 13 Okt 10 17:12 usb-Unknown_radino_CC1101-if00 -> ../../ttyACM0
Bootloader aktiv:
lrwxrwxrwx 1 root root 13 Okt 10 17:17 usb-In-Circuit_radino_CC1101-if00 -> ../../ttyACM0
Normalbetrieb:
lrwxrwxrwx 1 root root 13 Okt 10 17:17 usb-Unknown_radino_CC1101-if00 -> ../../ttyACM0

Reply:
Wozu?
Wenn das Device "usb-Unknown_radino_CC1101-if00" während der Bootloader 8 Sekunden aktiv ist, nicht existiert, nutzt auch längeres Warten nichts.
avrdude müsste während der Bootloader aktiv ist, auf diesen Port "usb-In-Circuit_radino_CC1101-if00" umschwenken.

Die Arduino-Software sucht z.B. selbst nach dem anderen Port. Unter Windows wird da z.B. aus COM3 dann COM4 oder ähnlich.